man1/projects.1
.\" Generated with Ronnjs 0.3.8
.\" http://github.com/kapouer/ronnjs/
.
.TH "FHC\-PROJECTS" "1" "June 2014" "" ""
.
.SH "NAME"
\fBfhc-projects\fR \-\- List FeedHenry Proejcts
.
.SH "SYNOPSIS"
.
.nf
fhc projects [list]
fhc projects create <project\-title> [<template\-id>]
fhc projects update <project\-id> <prop\-name> <value>
fhc projects read <project\-id>
fhc projects delete <project\-id>
fhc projects clone <project\-id>
    Note \'clone\' does a \'git clone\' of each each App in your Project into the current working directory\.
.
.fi
.
.SH "DESCRIPTION"
Access your FeedHenry Projects
.
.SS "list"
.
.nf
fhc projects [list]
.
.fi
.
.P
This command is the default option and can be used to list your FeedHenry Projects\. As with any command in FHC, passing a \'\-\-json\' flag will return the list of FeedHenry Apps in JSON format\.
.
.P
You can also specify a \'\-\-bare\' flag to the \'apps\' command, e\.g\. just show app guids:
.
.IP "" 4
.
.nf
$ fhc projects \-\-bare
t2Xva0\-4X2AFsxbBzsRLSet1
VCtqRZTJWnhz5b3SbcS1LvVm
DOK1rO5L2WAW7fjE_MAIehio
t\-Bq8rCXHMwJSY9heXZZs8AL
.
.fi
.
.IP "" 0
.
.P
show Project ids and titles only:
.
.IP "" 4
.
.nf
$ fhc apps \-\-bare="guid title"
t2Xva0\-4X2AFsxbBzsRLSet1 Project1
VCtqRZTJWnhz5b3SbcS1LvVm Project2
DOK1rO5L2WAW7fjE_MAIehio Project3
t\-Bq8rCXHMwJSY9heXZZs8AL Project4
.
.fi
.
.IP "" 0
.
.SS "create"
.
.nf
fhc projects create <project\-title> [<template\-id>]
.
.fi
.
.P
  Create is used to create a new FeedHenry Project based on a template\.  The templates available can be listed using the \fBfhc templates projects\fR command
.
.IP "" 4
.
.nf
fhc projects create \'FHC hello create\' hello_world_project
.
.fi
.
.IP "" 0
.
.P
  If no template\-id is specified  the default of \fBbare_project\fR will be used, which will create a project containing no apps\.
.
.SS "update"
.
.nf
fhc projects update <project\-id> <prop\-name> <value>
.
.fi
.
.P
  The only \fBprop\-name\fR supported for updating is \fBtitle\fR\|\.  The title can be updated using a command like:
.
.IP "" 4
.
.nf
fhc projects update DOK1rO5L2WAW7fjE_MAIehio title \'My New Project Title\'
.
.fi
.
.IP "" 0
.
.SS "read"
.
.nf
fhc projects read <project\-id>
Reads the information about the project specified, including the apps contained within it
.
.fi
.
.SS "delete"
.
.nf
fhc projects delete <project\-id>
Deletes the specified project
.
.fi
.
.SS "clone"
.
.nf
fhc projects clone <project\-id>
.
.fi
.
.P
  This clones each App in your Project locally\.  It does a \fBgit clone\fR of each each App in your Project into the current working directory,
  and needs the \fBgit\fR command to be installed locally and on your current execution path\. You must also have your SSH key uploaded to the FeedHenry platform\.


lib/projects.js
module.exports = projects;

projects.usage = "fhc projects [list]"
    + "\nfhc projects create <project-title> [<template-id>]"
    + "\nfhc projects update <project-id> <prop-name> <value>"
    + "\nfhc projects read <project-id>"
    + "\nfhc projects delete <project-id>"
    + "\nfhc projects clone <project-id>"
    + "\nNote 'clone' does a 'git clone' of each each App in your Project into the current working directory.";

var log = require("./utils/log");
var common = require("./common");
var fhreq = require("./utils/request");
var fhc = require("./fhc");
var ini = require('./utils/ini');
var _ = require('underscore');
var templates = require('./templates.js');
var util = require('util');
var fs = require('fs');
var clone = require('./clone.js');
var async = require('async');

function unknown(message, cb) {
  return cb(message + "\n" + "Usage: \n" + projects.usage);
}

function projects(args, cb) {
  if (args.length === 0) return listProjects(args, cb);

  var action = args[0];
  if ("list" === action) {
    return listProjects(args, cb);
  } else if ("create" === action) {
    return createProject(args, cb);
  } else if ("update" === action) {
    return updateProject(args, cb);
  } else if ("delete" === action) {
    return deleteProject(args, cb);
  } else if ("read" === action) {
    if (args.length !== 2) return cb(projects.usage);
    return readProject(args[1], cb);
  } else if ("clone" === action) {
    if (args.length !== 2) return cb(projects.usage);
    return cloneProject(args[1], cb);
  }else {
    return unknown("Invalid action: " + action, cb);
  }
}

function listProjects(args, cb) {
  common.listProjects(function (err, projs) {
    if (err) return cb(err);
    if (ini.get('table') === true) {
      projects.table = common.createTableForProjects(projs);
    }

    if(ini.get('bare') !== false) {
      var props = ['guid'];
      if (typeof ini.get('bare') === 'string') {
        props = ini.get('bare').split(" ");
      }
      projects.bare = '';
      _.each(projs, function(proj) {
        if (projects.bare !== '') projects.bare = projects.bare + '\n';
        for (var i=0; i<props.length; i++) {
          projects.bare = projects.bare + proj[props[i]] + " ";
        }
      });
    }

    return cb(err, projs);
  });
}

function createProject(args, cb) {
  if (args.length < 2) {
    return unknown("Invalid arguments", cb);
  }

  var title = args[1];
  var templateId = args[2] || "bare_project";
  var payload =  {
    title: title,
    apps:[],
    services:[]
  };

  function doCall(template, cb) {
    payload.template = template;
    common.doApiCall(fhreq.getFeedHenryUrl(), "box/api/projects", payload, "Error creating project: ", cb);
  }


  // Hackity hack: fh-art requires a custom blank template (which has no apps)
  // to facilitate this we have a sneaky 3'rd param here where we pass the template object in directly
  // This would be much cleaner if there was an API for creating Project Templates
  if (args[3] && (typeof args[3] === 'object')) {
    doCall(args[3], cb);
  } else {
    templates(['projects', templateId], function(err, template) {
      if (err) return cb(err);
      if (!template) return cb('Template not found: ' + templateId);
      doCall(template, cb);
    });
  }
}

function updateProject(args, cb) {
  if (args.length < 4) {
    return unknown("Invalid arguments", cb);
  }
  var projectId = args[1];
  var propName = args[2];
  var value = args[3];

  readProject(projectId, function(err, project) {
    if (err) return cb(err);
    project[propName] = value;
    fhreq.PUT(fhreq.getFeedHenryUrl(), "box/api/projects/" + projectId, project, function (err, remoteData, raw, response) {
      if (err) return cb(err);
      if (response.statusCode !== 200) return cb(raw);
      return cb(null, remoteData);
    });
  });
}

function readProject(projectId, cb) {
  common.doGetApiCall(fhreq.getFeedHenryUrl(), "box/api/projects/" + projectId, "Error reading Project: ", cb);
}

function deleteProject(args, cb) {
  if (args.length < 2) {
    return unknown("Invalid arguments", cb);
  }
  var endpoint = "box/api/projects/" + fhc.appId(args[1]);
  common.doDeleteApiCall(fhreq.getFeedHenryUrl(), endpoint, {},  "Error deleting project: ", function (err, data) {
    if (err) return cb(err);
    return cb(err, data);
  });
}

function cloneProject(projectId, cb) {
  common.listApps(projectId, function(err, apps) {
    if (err) return cb(err);
    function doClone(app, cb1) {
      clone([projectId, app.guid], cb1);
    }
    async.mapSeries(apps.list, doClone, function(err, results) {
      if (err) return cb(err);
      projects.message = '';
      _.each(results, function(result) {
        projects.message += result.stdout + '\n' + result.stderr;
      });
      return cb(null, results);
    });
  });
}

// bash completion
projects.completion = function (opts, cb) {
  var argv = opts.conf.argv.remain;
  if (argv[1] !== "projects") argv.unshift("projects");
  if (argv.length === 2) {
    return cb(null, ["clone", "create", "update", "read", "delete", "list"]);
  }

  if (argv.length === 3) {
    var action = argv[2];
    switch (action) {
      case "read":
      case "clone":
      case "update":
      case "delete":
        common.getProjectIds(cb);
        break;
      default: return cb(null, []);
    }
  }
}